### DISCUSSION TITLE
Accepts empty payload on add document routes

### DISCUSSION THREAD

**Discussion by gmourier** [meilisearch/product#319] [Feedback & Feature Proposal]
# Description

Since the `v0.24.0`, MeiliSearch raises an error when no documents are sent on `POST`/`PUT` Add Documents routes. e.g. `[]`

Although the main idea is to guide developers more easily to understand that they are not sending anything concrete on the way to add documents especially when the response is asynchronous but on the other side it can be verbose in the case the developers implement an indexing pipeline that might receive messages with an empty payload. 

It forces them to check if the payload of documents is empty before calling the method to avoid raising/silently catching the [error](https://github.com/meilisearch/specifications/blob/main/text/0061-error-format-and-definitions.md#variant-sending-a-valid-but-empty--payload-on-postput---indexesindexuiddocuments).

We would like to have your opinion on what you find the most coherent.

👍 Keep this error when the call is made without any documents in the payload.
👎 Remove this error and allow the call to this route without any documents in the payload.

**Comment by curquiza** [comment]
I'm not okay with an error, but maybe a warning 😇 
One day, maybe when MeiliSearch will handle warnings

**Reply by gmourier** [reply]
Thanks, @curquiza. Why are you not okay with an error? 😇

**Reply by curquiza** [reply]
Because it would stop my import with an error instead of just continuing, and for me it's not an error. I might want  to implement a cron that fetches some potenial-existing data every day to push them to MeiliSearch. In this case, I would rather not have any error making my cron job fails, but just the <s>update</s> task status saying "zero document imported"
Just my opinion

**Reply by gmourier** [reply]
Thank you! 

What worries me is to have beginners who might not index anything at all at the end and have to look for what could have happened by going to the documentation, finding, using the `task` endpoint to understand that they forgot to send documents.

Whereas the case you present here, in my opinion, is perhaps aimed at more experienced users who want to synchronize MeilISearch with other remote systems. It can also be facilitated by dedicated integrations that precisely manage this business use case before sending an empty payload to MeiliSearch.

I am an advocate of a robust API to guide users as much as possible to quickly reach the aha moment (doing a qualified, performant, and fast search) or the desired expectation but this is only my opinion too.

We can naturally think that someone who decides to pipeline documents to MeiliSearch would have already indexed documents and made searches to integrate MeiliSearch in its technical workflow. This is pure gut feeling but who buys before trying?

**Comment by gmourier** [comment]
I have a few questions to try to go further

- Why would you want to add documents but not know that you are sending nothing and thus found that 0 documents are indexed after all?
- If you don't send any documents and an error is not raised. Do you expect that MeiliSearch creates the index in a lazy way if it does not exist yet?

**Reply by irevoire** [reply]
For me, the obvious case of indexing zero documents is when you batch all your update and then send it at the end of the hour / day. You don't know if you did put something in your batch and don't want to check, Meilisearch should just handle that case for you (and shouldn't throw an error because now you have another check to do).

And yes, Meilisearch should create the index.

**Reply by MarinPostma** [reply]
In the case of a empty CSV addition, should the primary key be infered from the header?

**Reply by irevoire** [reply]
Yeah, that would be the expected behaviour to me

**Reply by MarinPostma** [reply]
so if you add no documents, and the primary key is not present in the header, this is an error?

**Comment by gmourier** [comment]
Lock since implemented in v0.26.0
